Fix pytest collection failure for classes decorated with context managers

kaxil opened the pull request:
Classes decorated with `@conf_vars` and other context managers were disappearing during pytest collection, causing tests to be silently skipped. This affected several test classes including `TestWorkerStart` in the Celery provider tests.

Root cause: `ContextDecorator` transforms decorated classes into callable wrappers. Since pytest only collects actual type objects as test classes, these wrapped classes are ignored during collection.

Simple reproduction (no Airflow needed):

```py
import contextlib
import inspect

@contextlib.contextmanager
def simple_cm():
    yield

@simple_cm()
class TestExample:
    def test_method(self):
        pass

print(f'Is class? {inspect.isclass(TestExample)}')  # False - pytest won't collect
```

and then run

```shell
pytest test_example.py --collect-only
```

Airflow reproduction:

**Before**
```shell
> breeze run pytest providers/celery/tests/unit/celery/cli/test_celery_command.py --collect-only -v

collected 6 items

<Dir airflow>
  <Dir providers>
    <Dir celery>
      <Dir tests>
        <Package unit>
          <Package celery>
            <Package cli>
              <Module test_celery_command.py>
                <Class TestRemoteCeleryControlCommands>
                  <Function test_list_celery_workers>
                  <Function test_shutdown_worker>
                  <Function test_shutdown_all_workers>
                  <Function test_add_queue>
                  <Function test_remove_queue>
                <Function test_stale_bundle_cleanup>
```

**After**:
```shell
> breeze run pytest providers/celery/tests/unit/celery/cli/test_celery_command.py --collect-only -v

collected 14 items

<Dir airflow>
  <Dir providers>
    <Dir celery>
      <Dir tests>
        <Package unit>
          <Package celery>
            <Package cli>
              <Module test_celery_command.py>
                <Class TestCeleryStopCommand>
                  <Function test_if_right_pid_is_read>
                  <Function test_same_pid_file_is_used_in_start_and_stop>
                  <Function test_custom_pid_file_is_used_in_start_and_stop>
                <Class TestWorkerStart>
                  <Function test_worker_started_with_required_arguments>
                <Class TestWorkerFailure>
                  <Function test_worker_failure_gracefull_shutdown>
                <Class TestFlowerCommand>
                  <Function test_run_command>
                  <Function test_run_command_daemon_v_3_below>
                  <Function test_run_command_daemon_v3_above>
                <Class TestRemoteCeleryControlCommands>
                  <Function test_list_celery_workers>
                  <Function test_shutdown_worker>
                  <Function test_shutdown_all_workers>
                  <Function test_add_queue>
                  <Function test_remove_queue>
                <Function test_stale_bundle_cleanup>
```

Solution:
1. Fixed affected test files by replacing class-level `@conf_vars` decorators with pytest fixtures
2. Created pytest fixtures to apply configuration changes
3. Used `@pytest.mark.usefixtures` to apply configuration to test classes
4. Added custom linter to prevent future occurrences and integrated it into pre-commit hooks

Files changed:
- Fixed 3 test files with problematic class decorators
- Added custom linter with pre-commit integration

This ensures pytest properly collects all test classes and prevents similar issues in the future through automated detection.

kaxil committed:
Fix pytest collection failure for classes decorated with context managers

Classes decorated with `@conf_vars` and other context managers were disappearing
during pytest collection, causing tests to be silently skipped. This affected
several test classes including `TestWorkerStart` in the Celery provider tests.

Root cause: `ContextDecorator` transforms decorated classes into callable wrappers.
Since pytest only collects actual type objects as test classes, these wrapped
classes are ignored during collection.

Simple reproduction (no Airflow needed):

```py
import contextlib
import inspect

@contextlib.contextmanager
def simple_cm():
    yield

@simple_cm()
class TestExample:
    def test_method(self):
        pass

print(f'Is class? {inspect.isclass(TestExample)}')  # False - pytest won't collect
```

and then run

```shell
pytest test_example.py --collect-only
```

Airflow reproduction:

```shell
breeze run pytest providers/celery/tests/unit/celery/cli/test_celery_command.py --collect-only -v

breeze run pytest providers/celery/tests/unit/celery/cli/test_celery_command.py --collect-only -v
```

Solution:
1. Fixed affected test files by replacing class-level `@conf_vars` decorators
   with pytest fixtures
2. Created pytest fixtures to apply configuration changes
3. Used `@pytest.mark.usefixtures` to apply configuration to test classes
4. Added custom linter to prevent future occurrences and integrated it
   into pre-commit hooks

Files changed:
- Fixed 3 test files with problematic class decorators
- Added custom linter with pre-commit integration

This ensures pytest properly collects all test classes and prevents similar
issues in the future through automated detection.

jedcunningham commented on scripts/ci/prek/check_contextmanager_class_decorators.py:
```diff
@@ -0,0 +1,131 @@
+                return decorator.func.id
+            if isinstance(decorator.func, ast.Attribute):
+                return f"{self._get_attr_chain(decorator.func)}"
+        elif isinstance(decorator, ast.Attribute):
```
```suggestion
        if isinstance(decorator, ast.Attribute):
```

?

jedcunningham commented on scripts/ci/prek/check_contextmanager_class_decorators.py:
```diff
@@ -0,0 +1,131 @@
+            "env_vars",
+            "contextlib.contextmanager",
+            "contextmanager",
+            # Add other known problematic decorators here
```
```suggestion
```

jedcunningham commented on scripts/ci/prek/check_contextmanager_class_decorators.py:
```diff
@@ -0,0 +1,131 @@
+        path = Path(arg)
+        if path.is_file() and path.suffix == ".py":
+            all_errors.extend(check_file(path))
+        elif path.is_dir():
```
prek only passes filenames yeah? This just for manual use?

kaxil replied:
Yeah, I used it to verify airflow-core, task-sdk & providers

jedcunningham commented on scripts/ci/prek/check_contextmanager_class_decorators.py:
```diff
@@ -0,0 +1,131 @@
+
+    for arg in sys.argv[1:]:
+        path = Path(arg)
+        if path.is_file() and path.suffix == ".py":
```
Should we want or fail if we get non-test files passed? Maybe also check for "test" here too?

kaxil replied:
Fixed

kaxil committed:
Fix linter logic and Kafka test fixtures

- Fix linter to consistently check only test files when processing individual files
- Add mock_supervisor_comms fixture to Kafka integration tests to resolve SUPERVISOR_COMMS import errors

The Kafka tests were failing because they execute operators directly via operator.execute()
which requires the task execution environment to be properly mocked.
